{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e4389de-1d6b-4df4-81a8-d419a356dfaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import linear_model\n",
    "import numpy as np\n",
    "from scipy.stats import pearsonr\n",
    "import seaborn as sns\n",
    "import datetime\n",
    "import requests\n",
    "from localcider.sequenceParameters import SequenceParameters\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from scipy import stats\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.decomposition import PCA\n",
    "import statsmodels.stats.multitest\n",
    "import sys\n",
    "import matplotlib as mpl\n",
    "from scipy.stats import linregress\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "from sklearn import linear_model\n",
    "import random\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from Bio.Align import PairwiseAligner\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram\n",
    "from Bio.Align import PairwiseAligner\n",
    "from Bio import pairwise2\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.metrics.pairwise import cosine_similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4eed4445-df1e-43e5-ba8a-eb296f3591f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = 'RBPs.xlsx'  # Replace with the path to your file\n",
    "df = pd.read_excel(file_path, header=0, skiprows=[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3172630a-dde4-4263-bdda-e527f83a1d60",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\faris\\AppData\\Local\\Temp\\ipykernel_23996\\4027190928.py:1: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  genemap_df = pd.read_csv('hgnc_complete_set_2024-10-01.txt', delimiter='\\t')\n"
     ]
    }
   ],
   "source": [
    "genemap_df = pd.read_csv('hgnc_complete_set_2024-10-01.txt', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc7d1054-be75-4040-bf98-2c3ae49b23c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_domain_annotations(uniprot_id):\n",
    "    \"\"\"Fetches domain annotations using the UniProt ID from the UniProt's features endpoint.\"\"\"\n",
    "    base_url = f\"https://www.ebi.ac.uk/proteins/api/features/{uniprot_id}\"\n",
    "    response = requests.get(base_url, headers={\"Accept\": \"application/json\"})\n",
    "    if response.ok:\n",
    "        domains = {}\n",
    "        for feat in response.json()['features']:\n",
    "            if feat['type'] == 'DOMAIN':\n",
    "                domains[ feat['description'] ] = {}\n",
    "                domains[ feat['description'] ]['begin'] = feat['begin']\n",
    "                domains[ feat['description'] ]['end'] = feat['end']\n",
    "                #print( feat ) \n",
    "                #feat['description']\n",
    "                #feat['begin']\n",
    "                #feat['end']\n",
    "        return domains\n",
    "#        return response.json()\n",
    "    \n",
    "    else:\n",
    "        raise ValueError(f\"Error fetching domain annotations for {uniprot_id}\")\n",
    "        \n",
    "\n",
    "def get_protein_sequence_mapped(gene_name ):\n",
    "    \n",
    "    uniprot_ids = genemap_df[genemap_df['ensembl_gene_id']==gene_name]['uniprot_ids'].tolist()\n",
    "\n",
    "    if len( uniprot_ids ) < 1: \n",
    "        print( gene_name )\n",
    "        return '', ''\n",
    "        \n",
    "    uniprot_id = uniprot_ids[0]\n",
    "    # URL to retrieve the sequence in FASTA format\n",
    "    url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "    \n",
    "    response = requests.get(url)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        fasta_data = response.text\n",
    "        # Parse the FASTA format and extract the sequence\n",
    "        lines = fasta_data.splitlines()\n",
    "        protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "        return protein_sequence, uniprot_id\n",
    "    else:\n",
    "        raise ValueError(f\"Protein sequence not found for UniProt ID: {uniprot_id}\")\n",
    "\n",
    "\n",
    "gene_name_to_protein_seq = {}\n",
    "gene_name_to_uniprot_id = {}\n",
    "gene_name_to_domains = {}\n",
    "for index, row in df.iterrows():\n",
    "    gene_name = row['Name']\n",
    "    ensembl_gene_id = row['ID']\n",
    "    ensembl_gene_id_map = ''\n",
    "    if ';' in ensembl_gene_id:\n",
    "        for gid in ensembl_gene_id.split(';'):\n",
    "            if gid in genemap_df['ensembl_gene_id'].tolist():\n",
    "                ensembl_gene_id_map = gid\n",
    "                break\n",
    "                \n",
    "    else:\n",
    "        ensembl_gene_id_map = ensembl_gene_id\n",
    "\n",
    "    if ensembl_gene_id_map == '':\n",
    "        print( f\"didn't find mapped id for {row['ID']}\" )\n",
    "    \n",
    "    try:\n",
    "        protein_sequence, uniprot_id = get_protein_sequence_mapped(gene_name=ensembl_gene_id_map)\n",
    "        gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "        gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "        if uniprot_id == '':\n",
    "            gene_name_to_domains[ gene_name ] = {}\n",
    "        else:\n",
    "            domains = get_domain_annotations( uniprot_id )\n",
    "            gene_name_to_domains[ gene_name ] = domains\n",
    "    except ValueError as e:\n",
    "        print(f\"Error for Gene: {gene_name}: {e}\")\n",
    "        \n",
    "    if (index % 100) == 0: \n",
    "        print( index )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70e9497e-8b70-4835-b40c-3f4b396f271b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually add the missing proteins\n",
    "uniprot_id = 'P42771'\n",
    "gene_name = 'CDKN2A'\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    fasta_data = response.text\n",
    "    # Parse the FASTA format and extract the sequence\n",
    "    lines = fasta_data.splitlines()\n",
    "    protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "domains = get_domain_annotations( uniprot_id )\n",
    "gene_name_to_domains[ gene_name ] = domains\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'Q9UPN3'\n",
    "gene_name = 'MACF1'\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    fasta_data = response.text\n",
    "    # Parse the FASTA format and extract the sequence\n",
    "    lines = fasta_data.splitlines()\n",
    "    protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "domains = get_domain_annotations( uniprot_id )\n",
    "gene_name_to_domains[ gene_name ] = domains\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'Q92614'\n",
    "gene_name = 'MYO18A'\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    fasta_data = response.text\n",
    "    # Parse the FASTA format and extract the sequence\n",
    "    lines = fasta_data.splitlines()\n",
    "    protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "domains = get_domain_annotations( uniprot_id )\n",
    "gene_name_to_domains[ gene_name ] = domains\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'P98175'\n",
    "gene_name = 'RBM10'\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    fasta_data = response.text\n",
    "    # Parse the FASTA format and extract the sequence\n",
    "    lines = fasta_data.splitlines()\n",
    "    protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "domains = get_domain_annotations( uniprot_id )\n",
    "gene_name_to_domains[ gene_name ] = domains\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'A0A0A6YYI9' # I think ?\n",
    "gene_name = 'RBM14-RBM4'\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    fasta_data = response.text\n",
    "    # Parse the FASTA format and extract the sequence\n",
    "    lines = fasta_data.splitlines()\n",
    "    protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "domains = get_domain_annotations( uniprot_id )\n",
    "gene_name_to_domains[ gene_name ] = domains\n",
    "\n",
    "################################################\n",
    "#omit RPL17-C18orf32 --> I am really not sure what protein this corresponds to\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'P42166' # the 2 are just different isoforms, not sure which is \"correct\", pick alpha for now\n",
    "gene_name = 'TMPO'\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    fasta_data = response.text\n",
    "    # Parse the FASTA format and extract the sequence\n",
    "    lines = fasta_data.splitlines()\n",
    "    protein_sequence = ''.join(lines[1:])  # Join all lines except the header\n",
    "gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "domains = get_domain_annotations( uniprot_id )\n",
    "gene_name_to_domains[ gene_name ] = domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "182913dc-7698-44e4-ac7c-c926446128bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['uniprot_id'] = df['Name'].map(gene_name_to_uniprot_id)\n",
    "df['sequence'] = df['Name'].map(gene_name_to_protein_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd026d0f-9b3f-44e9-b0bd-6263abc72f34",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def normalize_domain(domain_name):\n",
    "    # Removes trailing numbers and whitespace\n",
    "    return re.sub(r'\\s\\d+.*$', '', domain_name).strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206876da-1ac4-4fcd-a260-9b1cba6bb78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather domain names\n",
    "unique_domains = set(\n",
    "    normalize_domain(domain_name)\n",
    "    for domains in gene_name_to_domains.values()\n",
    "    for domain_name in domains\n",
    ")\n",
    "print(f\"Total unique domains: {len(unique_domains)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecfaf61a-7c5f-4005-b565-9ea009a764c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "genes = gene_name_to_domains.keys()\n",
    "sequences = gene_name_to_protein_seq.values()\n",
    "domain_data = defaultdict(dict)\n",
    "\n",
    "# Build stats per gene\n",
    "for gene, sequence in zip(genes, sequences):\n",
    "    seq_len = len(sequence)\n",
    "    domain_stats = defaultdict(list)\n",
    "    domain_string_map = defaultdict(lambda: [\"0\"] * seq_len)\n",
    "    \n",
    "    for raw_name, pos_dict in gene_name_to_domains[gene].items():\n",
    "        norm_name = normalize_domain(raw_name)\n",
    "        try:\n",
    "            begin = int(pos_dict['begin'])\n",
    "            end = int(pos_dict['end'])\n",
    "            length = abs(end - begin)\n",
    "            domain_stats[norm_name].append(length)\n",
    "\n",
    "            for i in range(begin, end):\n",
    "                    if 0 <= i < seq_len:\n",
    "                        domain_string_map[norm_name][i] = \"1\"\n",
    "                    \n",
    "        except (ValueError, KeyError):\n",
    "            continue\n",
    "    \n",
    "    for domain in unique_domains:\n",
    "        lengths = domain_stats.get(domain, [])\n",
    "        domain_data[gene][f\"Number of {domain}\"] = len(lengths)\n",
    "        domain_data[gene][f\"Average Size of {domain}\"] = sum(lengths)/len(lengths) if lengths else 0\n",
    "        domain_data[gene][f\"Total Size of {domain}\"] = sum(lengths)\n",
    "        domain_data[gene][f\"{domain} Map\"] = \"\".join(domain_string_map[domain])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a70be2a-7507-4791-9632-cb5a987efd7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert domain_data to a DataFrame\n",
    "domain_df = pd.DataFrame.from_dict(domain_data, orient=\"index\")\n",
    "\n",
    "# Merge with existing DataFrame\n",
    "df = df.merge(domain_df, left_on=\"Name\", right_index=True, how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e06dce72-4aeb-4082-8246-33eafbca1255",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_zinc_finger_annotations(uniprot_id):\n",
    "    \"\"\"Fetches Zinc Finger regions labeled 'ZN_FING' from UniProt's features API.\"\"\"\n",
    "    base_url = f\"https://www.ebi.ac.uk/proteins/api/features/{uniprot_id}\"\n",
    "    response = requests.get(base_url, headers={\"Accept\": \"application/json\"})\n",
    "\n",
    "    if response.ok:\n",
    "        zinc_features = {}\n",
    "        for feat in response.json().get(\"features\", []):\n",
    "            if feat.get(\"type\") == \"ZN_FING\":\n",
    "                desc = feat.get(\"description\", \"Zinc Finger\")\n",
    "                zinc_features[desc] = {\n",
    "                    \"begin\": feat[\"begin\"],\n",
    "                    \"end\": feat[\"end\"]\n",
    "                }\n",
    "        return zinc_features\n",
    "    else:\n",
    "        raise ValueError(f\"Failed to fetch features for UniProt ID {uniprot_id}\")\n",
    "\n",
    "gene_name_to_fingers = {}\n",
    "for index, row in df.iterrows():\n",
    "    #gene_name = row['UNIQUE']  # Assuming \"UNIQUE\" is the gene name column\n",
    "    #print( index, row['ID'] )\n",
    "    gene_name = row['Name']\n",
    "    ensembl_gene_id = row['ID'] # Assuming \"ID\" is the gene ID column\n",
    "    #if ensembl_gene_id != 'ENSG00000231135': continue\n",
    "    ensembl_gene_id_map = ''\n",
    "    if ';' in ensembl_gene_id:\n",
    "        for gid in ensembl_gene_id.split(';'):\n",
    "            if gid in genemap_df['ensembl_gene_id'].tolist():\n",
    "                ensembl_gene_id_map = gid\n",
    "                break\n",
    "                \n",
    "    else:\n",
    "        ensembl_gene_id_map = ensembl_gene_id\n",
    "\n",
    "    if ensembl_gene_id_map == '':\n",
    "        print( f\"didn't find mapped id for {row['ID']}\" )\n",
    "    \n",
    "    try:\n",
    "        protein_sequence, uniprot_id = get_protein_sequence_mapped(gene_name=ensembl_gene_id_map)\n",
    "        gene_name_to_protein_seq[ gene_name ] = protein_sequence\n",
    "        gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "        if uniprot_id == '':\n",
    "            gene_name_to_fingers[ gene_name ] = {}\n",
    "        else:\n",
    "            fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "            gene_name_to_fingers[ gene_name ] = fingers\n",
    "        #print(f\"Gene: {gene_name}, Ensembl ID: {ensembl_gene_id}, Protein Sequence: {protein_sequence[:50]}...\")  # Print first 50 characters of the sequence\n",
    "    except ValueError as e:\n",
    "        print(f\"Error for Gene: {gene_name}: {e}\")\n",
    "        \n",
    "    if (index % 100) == 0: \n",
    "        print( index )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "654c9059-3e75-4c81-9b83-779db57fff73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually add the missing proteins\n",
    "uniprot_id = 'P42771'\n",
    "gene_name = 'CDKN2A'\n",
    "\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "gene_name_to_fingers[ gene_name ] = fingers\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'Q9UPN3'\n",
    "gene_name = 'MACF1'\n",
    "\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "gene_name_to_fingers[ gene_name ] = fingers\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'Q92614'\n",
    "gene_name = 'MYO18A'\n",
    "\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "gene_name_to_fingers[ gene_name ] = fingers\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'P98175'\n",
    "gene_name = 'RBM10'\n",
    "\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "gene_name_to_fingers[ gene_name ] = fingers\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'A0A0A6YYI9' # I think ?\n",
    "gene_name = 'RBM14-RBM4'\n",
    "\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "gene_name_to_fingers[ gene_name ] = fingers\n",
    "\n",
    "################################################\n",
    "#omit RPL17-C18orf32 --> I am really not sure what protein this corresponds to\n",
    "\n",
    "################################################\n",
    "uniprot_id = 'P42166' # the 2 are just different isoforms, not sure which is \"correct\", pick alpha for now\n",
    "gene_name = 'TMPO'\n",
    "\n",
    "gene_name_to_uniprot_id[ gene_name ] = uniprot_id\n",
    "fingers = get_zinc_finger_annotations( uniprot_id )\n",
    "gene_name_to_fingers[ gene_name ] = fingers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0829c9d-a575-47f0-ae99-1ff7fb3295a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather zinc finger names\n",
    "unique_fingers = set(\n",
    "    normalize_domain(finger_name)\n",
    "    for fingers in gene_name_to_fingers.values()\n",
    "    for finger_name in fingers\n",
    ")\n",
    "\n",
    "print(f\"Total unique Zinc Fingers: {len(unique_fingers)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8ce591e-2fc9-4950-b8a3-d9d55438d1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = gene_name_to_domains.keys()\n",
    "sequences = gene_name_to_protein_seq.values()\n",
    "finger_data = defaultdict(dict)\n",
    "\n",
    "# Build stats per gene\n",
    "for gene, sequence in zip(genes, sequences):\n",
    "    seq_len = len(sequence)\n",
    "    finger_stats = defaultdict(list)\n",
    "    finger_string_map = defaultdict(lambda: [\"0\"] * seq_len)\n",
    "    \n",
    "    for raw_name, pos_dict in gene_name_to_fingers[gene].items():\n",
    "        norm_name = normalize_domain(raw_name)\n",
    "        try:\n",
    "            begin = int(pos_dict['begin'])\n",
    "            end = int(pos_dict['end'])\n",
    "            length = abs(end - begin)\n",
    "            finger_stats[norm_name].append(length)\n",
    "\n",
    "            for i in range(begin, end):\n",
    "                    if 0 <= i < seq_len:\n",
    "                        finger_string_map[norm_name][i] = \"1\"\n",
    "                    \n",
    "        except (ValueError, KeyError):\n",
    "            continue\n",
    "    \n",
    "    for finger in all_fingers:\n",
    "        lengths = finger_stats.get(finger, [])\n",
    "        finger_data[gene][f\"Number of {finger}\"] = len(lengths)\n",
    "        finger_data[gene][f\"Average Size of {finger}\"] = sum(lengths)/len(lengths) if lengths else 0\n",
    "        finger_data[gene][f\"Total Size of {finger}\"] = sum(lengths)\n",
    "        finger_data[gene][f\"{finger} Map\"] = \"\".join(finger_string_map[finger])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14494097-46b1-45c9-9196-004b41a8fbef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert finger_data to a DataFrame\n",
    "finger_df = pd.DataFrame.from_dict(finger_data, orient=\"index\")\n",
    "\n",
    "# Merge with existing DataFrame\n",
    "df = df.merge(finger_df, left_on=\"Name\", right_index=True, how=\"left\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
